src/boot/fwupd.c

```diff
@@ -0,0 +1,630 @@
+                                    *n_updates * sizeof(FwupUpdateTable),
+                                    (*n_updates + 1) * sizeof(FwupUpdateTable));
+                (*updates)[(*n_updates)++] = TAKE_PTR(update);
+        }
+
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS get_graphics_protocol_mode(EFI_HANDLE image, uint32_t *ret_mode) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        size_t n_handles = 0;
+        static uint32_t mode = 0;
+        static int found = -1;
+        EFI_STATUS err;
+
+        assert(ret_mode);
+
+        /* Doesn't depend on the capsule being inspected, cache the result */
+        if (found > 0) {
+                *ret_mode = mode;
+                return EFI_SUCCESS;
+        }
+        if (found == 0)
+                return EFI_UNSUPPORTED;
+
+        err = BS->LocateHandleBuffer(
+                        ByProtocol,
+                        MAKE_GUID_PTR(EFI_GRAPHICS_OUTPUT_PROTOCOL),
+                        /* SearchKey= */ NULL,
+                        &n_handles,
+                        &handles);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to locate graphics output protocol handles: %m");
+        if (n_handles == 0 || !handles)
+                return log_error_status(EFI_UNSUPPORTED, "No graphics output protocol handles found for UX capsule");
+
+        FOREACH_ARRAY(h, handles, n_handles) {
+                EFI_GRAPHICS_OUTPUT_PROTOCOL *proto;
+
+                err = BS->OpenProtocol(*h,
+                                       MAKE_GUID_PTR(EFI_GRAPHICS_OUTPUT_PROTOCOL),
+                                       (void **)&proto,
+                                       image,
+                                       /* ControllerHandler= */ NULL,
+                                       EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to get graphics output protocol: %m");
+
+                *ret_mode = mode = proto->mode->mode;
+                found = 1;
+                return EFI_SUCCESS;
+        }
+
+        found = 0;
+        return EFI_UNSUPPORTED;
+}
+
+static EFI_DEVICE_PATH *device_path_find_type(EFI_DEVICE_PATH **dp, uint8_t type) {
+        EFI_DEVICE_PATH *next;
+
+        assert(dp);
+        assert(*dp);
+
+        for (;;) {
+                next = device_path_next_node(*dp);
+
+                if ((*dp)->Type == type)
+                        break;
+
+                *dp = next;
+        }
+
+        return next;
+}
+
+static EFI_DEVICE_PATH *device_path_instance(EFI_DEVICE_PATH **device_path, size_t *ret_size) {
+        EFI_DEVICE_PATH *head, *dp;
+
+        assert(device_path);
+        assert(ret_size);
+
+        if (!*device_path) {
+                *ret_size = 0;
+                return NULL;
+        }
+
+        dp = head = *device_path;
+
+        /* Skip through all subtypes until the last node and return it */
```

Comment:
i cannot parse this comment.

Reply:
I'm not sure how else to phrase it?

Reply:
you are not skipping through "subtypes", you are skipping through nodes of a device path... are you sure you understand the function you are adding here yourself?

Reply:
If you have a better suggestion, please just write it down and I'll copy and paste it